# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/canvas/WebGLContextUtils.cpp
# Commit: 71c122ac0ca7
# Full Hash: 71c122ac0ca73391866b1ef19f4f82bc2d28568b
# Author: David Parks <davidp99@gmail.com>
# Date: 2020-01-09 03:40:18
# Regressor Bug: 1477756
# File Overlap Count: 1
# Description:
#   Bug 1477756 - Initial out-of-process WebGL implementation. r=mccr8,handyman
#   
#   Splits WebGLContext into ClientWebGLContext and HostWebGLContext.  The Client enables the JS-control of a WebGL context in a content procecss while the Host executes the WebGL graphics operations (via a WebGLContext that maintains much of the existing code) in the compositor process.  At this point, the cross-process behavior is disabled -- this series of patches is an incremental step toward that final goal.
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D54018
# ==============================================================================

diff -r 5cbf5bb11c58 -r 71c122ac0ca7 dom/canvas/WebGLContextUtils.cpp
--- a/dom/canvas/WebGLContextUtils.cpp	Wed Jan 08 21:52:03 2020 +0000
+++ b/dom/canvas/WebGLContextUtils.cpp	Wed Jan 08 22:19:14 2020 +0000
@@ -5,7 +5,7 @@
 
 #include "WebGLContextUtils.h"
 #include "WebGLContext.h"
-
+#include "HostWebGLContext.h"
 #include "GLContext.h"
 #include "jsapi.h"
 #include "js/Warnings.h"  // JS::WarnASCII
@@ -40,16 +40,6 @@
   }
 }
 
-JS::Value StringValue(JSContext* cx, const char* chars, ErrorResult& rv) {
-  JSString* str = JS_NewStringCopyZ(cx, chars);
-  if (!str) {
-    rv.Throw(NS_ERROR_OUT_OF_MEMORY);
-    return JS::NullValue();
-  }
-
-  return JS::StringValue(str);
-}
-
 void WebGLContext::GenerateWarning(const char* fmt, ...) const {
   va_list ap;
   va_start(ap, fmt);
@@ -68,25 +58,18 @@
   VsprintfLiteral(buf, fmt, ap);
 
   // JS::WarnASCII will print to stderr for us.
-
-  if (!mCanvasElement) {
-    return;
-  }
-
-  dom::AutoJSAPI api;
-  if (!api.Init(mCanvasElement->OwnerDoc()->GetScopeObject())) {
-    return;
+  const auto funcName = FuncName();
+  nsCString msg;
+  msg.AppendPrintf("WebGL warning: %s: %s", funcName, buf);
+  if (!ShouldGenerateWarnings()) {
+    msg.AppendPrintf(
+        "WebGL: No further warnings will be reported for"
+        " this WebGL context."
+        " (already reported %d warnings)",
+        mAlreadyGeneratedWarnings);
   }
 
-  JSContext* cx = api.cx();
-  const auto funcName = FuncName();
-  JS::WarnASCII(cx, "WebGL warning: %s: %s", funcName, buf);
-  if (!ShouldGenerateWarnings()) {
-    JS::WarnASCII(cx,
-                  "WebGL: No further warnings will be reported for this WebGL "
-                  "context. (already reported %d warnings)",
-                  mAlreadyGeneratedWarnings);
-  }
+  mHost->PostWarning(msg);
 }
 
 bool WebGLContext::ShouldGenerateWarnings() const {
@@ -98,14 +81,6 @@
 void WebGLContext::GeneratePerfWarning(const char* fmt, ...) const {
   if (!ShouldGeneratePerfWarnings()) return;
 
-  if (!mCanvasElement) return;
-
-  dom::AutoJSAPI api;
-  if (!api.Init(mCanvasElement->OwnerDoc()->GetScopeObject())) return;
-  JSContext* cx = api.cx();
-
-  ////
-
   va_list ap;
   va_start(ap, fmt);
 
@@ -117,15 +92,18 @@
   ////
 
   const auto funcName = FuncName();
-  JS::WarnASCII(cx, "WebGL perf warning: %s: %s", funcName, buf);
+  nsCString msg;
+  msg.AppendPrintf("WebGL perf warning: %s: %s", funcName, buf);
   mNumPerfWarnings++;
 
   if (!ShouldGeneratePerfWarnings()) {
-    JS::WarnASCII(cx,
-                  "WebGL: After reporting %u, no further perf warnings will be "
-                  "reported for this WebGL context.",
-                  uint32_t(mNumPerfWarnings));
+    msg.AppendPrintf(
+        "WebGL: After reporting %u, no further perf warnings will"
+        " be reported for this WebGL context.",
+        uint32_t(mNumPerfWarnings));
   }
+
+  mHost->PostWarning(msg);
 }
 
 void WebGLContext::SynthesizeGLError(GLenum err) const {
@@ -733,27 +711,27 @@
              int4[2] == mViewportWidth && int4[3] == mViewportHeight);
 
   AssertUintParamCorrect(gl, LOCAL_GL_PACK_ALIGNMENT,
-                         mPixelStore_PackAlignment);
+                         mPixelStore.mPackAlignment);
   AssertUintParamCorrect(gl, LOCAL_GL_UNPACK_ALIGNMENT,
-                         mPixelStore_UnpackAlignment);
+                         mPixelStore.mUnpackAlignment);
 
   if (IsWebGL2()) {
     AssertUintParamCorrect(gl, LOCAL_GL_UNPACK_IMAGE_HEIGHT,
-                           mPixelStore_UnpackImageHeight);
+                           mPixelStore.mUnpackImageHeight);
     AssertUintParamCorrect(gl, LOCAL_GL_UNPACK_SKIP_IMAGES,
-                           mPixelStore_UnpackSkipImages);
+                           mPixelStore.mUnpackSkipImages);
     AssertUintParamCorrect(gl, LOCAL_GL_UNPACK_ROW_LENGTH,
-                           mPixelStore_UnpackRowLength);
+                           mPixelStore.mUnpackRowLength);
     AssertUintParamCorrect(gl, LOCAL_GL_UNPACK_SKIP_ROWS,
-                           mPixelStore_UnpackSkipRows);
+                           mPixelStore.mUnpackSkipRows);
     AssertUintParamCorrect(gl, LOCAL_GL_UNPACK_SKIP_PIXELS,
-                           mPixelStore_UnpackSkipPixels);
+                           mPixelStore.mUnpackSkipPixels);
     AssertUintParamCorrect(gl, LOCAL_GL_PACK_ROW_LENGTH,
-                           mPixelStore_PackRowLength);
+                           mPixelStore.mPackRowLength);
     AssertUintParamCorrect(gl, LOCAL_GL_PACK_SKIP_ROWS,
-                           mPixelStore_PackSkipRows);
+                           mPixelStore.mPackSkipRows);
     AssertUintParamCorrect(gl, LOCAL_GL_PACK_SKIP_PIXELS,
-                           mPixelStore_PackSkipPixels);
+                           mPixelStore.mPackSkipPixels);
   }
 
   MOZ_ASSERT(!gl::GLContext::IsBadCallError(errorScope.GetError()));
@@ -799,14 +777,4 @@
 
 ////
 
-JS::Value StringValue(JSContext* cx, const nsAString& str, ErrorResult& er) {
-  JSString* jsStr = JS_NewUCStringCopyN(cx, str.BeginReading(), str.Length());
-  if (!jsStr) {
-    er.Throw(NS_ERROR_OUT_OF_MEMORY);
-    return JS::NullValue();
-  }
-
-  return JS::StringValue(jsStr);
-}
-
 }  // namespace mozilla